src/main/java/ch/njol/skript/events/EvtGrow.java

```diff
@@ -34,56 +39,188 @@ public class EvtGrow extends SkriptEvent {
+
+		switch (actionRestriction) {
+			case OF:
+				if (eventRestriction == STRUCTURE) {
+					fromTypes = (Literal<Object>) args[0];
+				} else if (eventRestriction == BLOCK) {
+					fromTypes = (Literal<Object>) args[1];
+				}
+				break;
+			case FROM:
+				fromTypes = (Literal<Object>) args[0];
+				break;
+			case INTO:
+				if (eventRestriction == STRUCTURE) {
+					toTypes = (Literal<Object>) args[0];
+				} else if (eventRestriction == BLOCK) {
+					toTypes = (Literal<Object>) args[1];
+				}
+				break;
+			case FROM_INTO:
+				fromTypes = (Literal<Object>) args[0];
+				if (eventRestriction == STRUCTURE) {
+					toTypes = (Literal<Object>) args[1];
+				} else if (eventRestriction == BLOCK) {
+					toTypes = (Literal<Object>) args[2];
+				}
+				break;
+			default:
+				assert false;
+				return false;
 		}
-		// Else: no type restrictions specified
 		return true;
 	}
 	
 	@Override
 	public boolean check(Event event) {
-		if (eventType == STRUCTURE && types != null && event instanceof StructureGrowEvent) {
-			return types.check(event, type -> type.is(((StructureGrowEvent) event).getSpecies()));
-		} else if (eventType == BLOCK && blocks != null && event instanceof BlockGrowEvent) {
-			return blocks.check(event, type -> type.isOfType(((BlockGrowEvent) event).getNewState()));
+		if (fromTypes == null && actionRestriction != INTO)
+			// True for "on grow", false for anything else.
+			return actionRestriction == OF;
+
+		// Can exit early if we're checking against a structure, but the event isn't a structure grow event
+		// Can also exit early if we're checking against a block, but the event isn't a block grow event AND we're not checking for OF
+		// With OF, we can have `on grow of sapling` or `big mushroom` be a StructureGrowEvent
+		if (eventRestriction == STRUCTURE && !(event instanceof StructureGrowEvent)) {
+			return false;
+		} else if (eventRestriction == BLOCK && !(event instanceof BlockGrowEvent) && actionRestriction != OF) {
+			return false;
+		}
+
+		switch (actionRestriction) {
+			case OF:
+				return checkFrom(event, fromTypes) || checkTo(event, fromTypes);
+			case FROM:
+				return checkFrom(event, fromTypes);
+			case INTO:
+				return checkTo(event, toTypes);
+			case FROM_INTO:
+				return checkFrom(event, fromTypes) && checkTo(event, toTypes);
+			default:
+				assert false;
+				return false;
 		}
-		return event instanceof BlockGrowEvent || event instanceof StructureGrowEvent;
 	}
-	
+
+	private static boolean checkFrom(Event event, Literal<Object> types) {
+		if (types instanceof LiteralList)
+			((LiteralList<Object>) types).setAnd(false);
+		if (event instanceof StructureGrowEvent) {
+			// We want the original block, not the new one, so we get the first blockstate in the list and get its block
+			// Feels a bit hacky, but it seems to be the only way to get the original block
+			Block rootBlock = ((StructureGrowEvent) event).getBlocks().get(0).getBlock();
```

Comment:
Was there any case this would return false positive? (I never touched strcture grow event so I actually have no idea what it usually return) if so would it be any better to check for all blocks in StructureGrowEvent#getBlocks()?

Reply:
In all the trees/mushrooms I tested, it worked fine. Though I have a sneaking suspicion mangroves may cause issues. I'll take a closer look.

Reply:
I'm relatively certain the first element will always be the block at the location, but I changed it to use getLocation as that seems safer.

Reply:
Sounds good.